Update flexible_event_config.md to include Lite Flexible Event Level … (#882)
WICG/attribution-reporting-api: flexible_event_config.md

  - [API Changes](#api-changes)
  - [Default Configurations](#default-configurations)
    - [Default event sources](#default-event-sources)
    - [Default navigation sources](#default-navigation-sources)
    - [Custom Configurations: Examples](#custom-configurations-examples)
      - [Reporting trigger value buckets](#reporting-trigger-value-buckets)
      - [Reporting trigger counts](#reporting-trigger-counts)
      - [Binary with more frequent reporting](#binary-with-more-frequent-reporting)
      - [Varying trigger_specs from source to source](#varying-trigger_specs-from-source-to-source)
- [Privacy Considerations](#privacy-considerations)

The default configuration for event and navigation sources may not be ideal for all use-cases. We can optionally support extended configurations that allow for callers to specify precisely the information they want out of reports, in order to more efficiently extract utility out of the privacy mechanism. The most efficient configuration will differ from use-case to use-case and will depend on a) the parameters of our privacy mechanism and b) the noise level that can be tolerated by the use-case.

> We plan to implement this proposal in two phases:

Reviewer: ```suggestion
This proposal is broken into two separate feature sets
```

* __Phase 1: Lite flexible event level configuration__

>   * This lite version would provide a subset of the full feature and can be used independently of the full version of flexible event level

Reviewer: ```suggestion
  * This lite version would provide a subset of the full feature and can be used independently of Phase 2
```

## Phase 1: Lite Flexible Event Level

### API Changes

We will add the following two optional parameters to the JSON in `Attribution-Reporting-Register-Source`: `max_event_level_reports` and `event_report_windows`:

```jsonc
{
  ...
  // Optional. This is a parameter that acts across all trigger types for the lifetime of this source. 
  // It restricts the total number of event-level reports that this source can generate. 
  // After this maximum is hit, the source is no longer capable of producing any new data. 
  // The use of priority in the trigger attribution algorithm in the case of multiple attributable triggers remains unchanged. 
  // Defaults to 3 for navigation sources and 1 for event sources
  "max_event_level_reports": <int>,

  // Optional. Represents a series of time windows, starting at <start time>.
  // Reports for this source will be delivered an hour after the end of each window.
  // Time is encoded as seconds after source registration.
  // If event_report_windows is omitted, will use the default windows.
  // End time is exclusive.

>   "event_report_windows": {

Reviewer 2: Do we also need to say that this field is mutually exclusive with the existing `event_report_window` field?

Author: updated the code comment

>     "start_time": <int>, // Optional, defaults to 0.

Reviewer 2: I thought this was no longer being proposed for inclusion in phase 1?

Author: I have removed "start_time"

    "end_times": [<int>, ...]
  }
}
```

Note that if both `event_report_window` and `event_report_windows` are present then the source will be ignored. Only one of them can be present in the source registration.

### Default Configurations
Here are the default configurations for `event` and `navigation` sources.

#### Default event sources

```jsonc
{
  ...  
  "max_event_level_reports": 1,
  "event_report_windows": {

>     "end_times": [<30 days>]

Reviewer 2: Maybe we can include the actual integer value corresponding to 30 days here, and use a comment to indicate that it is equal to 30 days. Same below.

>     "end_times": [2592000] // 30 days represented in seconds

Author: I have added in the event_report_windows in the phase 2 examples

  ...  
  "max_event_level_reports": 1,
  "event_report_windows": {

>     "end_times": [2592000] //This is <30 days> represented in seconds

Reviewer 2: ```suggestion
    "end_times": [2592000] // 30 days represented in seconds
```

  ...  
  "max_event_level_reports": 3,
  "event_report_windows": {

>     "end_times": [172800, 604800, 2592000] //This is <2 days>, <7 days>, <30 days> represented in seconds

Reviewer 2: ```suggestion
    "end_times": [172800, 604800, 2592000] // 2 days, 7 days, 30 days represented in seconds
```

  ...  
  "max_event_level_reports": 3,
  "event_report_windows": {

>     "end_times": [7200, 43200, 86400] //This is <2 hours>, <12 hours>, and <1 day> represented in seconds

Reviewer 2: ```suggestion
    "end_times": [7200, 43200, 86400] // 2 hours, 12 hours, 1 day represented in seconds
```

    // Reports for this spec will be delivered an hour after the end of each window.
    // Time is encoded as seconds after source registration.
    //
    // Note: specs with identical trigger_data cannot have overlapping windows;
    // this ensures that triggers match at most one spec. If event_report_windows

>     // is omitted, will use the "event_report_window" or "event_report_windows" field specified at the global level for the source (or the defaul windows if none are specified).  End time is exclusive.

Reviewer 2: ```suggestion
    // is omitted, will use the "event_report_window" or "event_report_windows" field specified at the global level for the source (or the default windows if none are specified).  End time is exclusive.
```

    //
    // If omitted, then represents a trivial mapping
    // [1, 2, ... , MAX_INT]
    // With MAX_INT being the maximum int value defined by the browser.
    "summary_buckets": [<bucket start>, ...]
  }, {
    // Next trigger_spec
  } ...],

>   // Part of phase 1. Optional. This is a global parameter that acts across all trigger specs

Reviewer: ditto to below

Author: updated code comments

  // for the lifetime of this source. It restricts the total number of event-level reports
  // that this source can generate. This parameter also restricts the total number of bucket increments as well by limiting the number of conversions per source.
  // After this maximum is hit, the source is no longer capable of producing any new
  // data. The use of priority in the trigger attribution algorithm in the case of
  // multiple attributable triggers remains unchanged.
  // Defaults to 3 for navigation sources and 1 for event sources.
  "max_event_level_reports": <int>

>   // Part of phase 1. Optional. Represents a series of time windows, starting at <start time>.

Reviewer: We can probably just say:

// See description in phase 1.

As this is a 1:1 copy above the prose above.

Author: updated code comments

  "max_event_level_reports": <int>
  // See description in phase 1.
  "event_report_windows": {

>     "end_times": [7200, 43200, 86400] //This is <2 hours>, <12 hours>, and <1 day> represented in seconds

Reviewer 2: ```suggestion
    "end_times": [7200, 43200, 86400] // 2 hours, 12 hours, 1 day represented in seconds
```

* Apply global attribution filters
* For every trigger spec:
  * Evaluate the `event_trigger_data` on the spec to find a match, using the spec’s `event_reporting_window`

>     * The global `event_reporting_windows` (from Phase 1) will act as a default value in case any trigger spec is the missing `event_report_windows` sub-field

Reviewer: ```suggestion
    * The top level `event_reporting_windows` will act as a default value in case any trigger spec is the missing `event_report_windows` sub-field
```

  {
    "trigger_data": [0, 1, 2, 3, 4, 5, 6, 7],
    "event_report_windows": {

>       "end_times": [172800, 604800, 2592000] //This is <2 days>, <7 days>, <30 days> represented in seconds

Reviewer 2: ```suggestion
      "end_times": [172800, 604800, 2592000] // 2 days, 7 days, 30 days represented in seconds
```

  {
    "trigger_data": [0],
    "event_report_windows": {

>       "end_times": [604800] //This is <7 days> represented in seconds

Reviewer 2: ```suggestion
      "end_times": [604800] // 7 days represented in seconds
```

  {
    "trigger_data": [0],
    "event_report_windows": {

>       "end_times": [604800] //This is <7 days> represented in seconds

Reviewer 2: ```suggestion
      "end_times": [604800] // 7 days represented in seconds
```

  {
    "trigger_data": [0],
    "event_report_windows": {

>       // This is <1 days>, <2 days>, <3 days> <5 days>, <7 days>, <10 days> represented in seconds

Reviewer 2: ```suggestion
      // 1 day, 2 days, 3 days, 5 days, 7 days, 10 days represented in seconds
```

  {
    "trigger_data": [0, 1, 2, 3],
    "event_report_windows": {

>       "end_times": [172800, 604800, 2592000] //This is <2 days>, <7 days>, <30 days> represented in seconds

Reviewer 2: ```suggestion
      "end_times": [172800, 604800, 2592000] // 2 days, 7 days, 30 days represented in seconds
```

  {
    "trigger_data": [4, 5, 6, 7],
    "event_report_windows": {

>       "end_times": [172800, 604800, 2592000] //This is <2 days>, <7 days>, <30 days> represented in seconds

Reviewer 2: ```suggestion
      "end_times": [172800, 604800, 2592000] // 2 days, 7 days, 30 days represented in seconds
```
